cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java

```diff
@@ -0,0 +1,804 @@
+        if (!dryrun) {
+            rebalanceOptionsBuilder.withFullRun();
+        }
+        return apiClient.rebalance(host, CruiseControl.REST_API_PORT, rebalanceOptionsBuilder.build(), userTaskID)
+                .map(response -> {
+                    if (dryrun) {
+                        if (response.thereIsNotEnoughDataForProposal() || response.proposalIsStillCalculating()) {
+                            // If either there not enough data for a rebalance or it is still being processed we need to
+                            // re-request the proposal at a later stage so we move to the PendingProposal State.
+                            return new KafkaClusterRebalanceStatusBuilder()
+                                    .withNewSessionId(response.getUserTaskId())
+                                    .addNewCondition().withNewType(State.PendingProposal.toString()).endCondition().build();
+                        }
+                    } else {
+                        if (response.thereIsNotEnoughDataForProposal()) {
+                            // We do not include a session id with this status as we do not want to retrieve the state of
+                            // this failed tasks (COMPLETED_WITH_ERROR)
+                            return new KafkaClusterRebalanceStatusBuilder()
+                                    .addNewCondition().withNewType(State.PendingProposal.toString()).endCondition().build();
+                        } else if (response.proposalIsStillCalculating()) {
+                            return new KafkaClusterRebalanceStatusBuilder()
+                                    .withNewSessionId(response.getUserTaskId())
+                                    .addNewCondition().withNewType(State.Rebalancing.toString()).endCondition().build();
+                        }
+                    }
+
+                    if (response.getJson().containsKey(CC_REST_API_SUMMARY)) {
+                        // If there is enough data and the proposal is complete (the response has the "summary" key) then we move
+                        // to ProposalReady for a dry run or to the Rebalancing state for a full run
+                        State ready = dryrun ? State.ProposalReady : State.Rebalancing;
+                        return new KafkaClusterRebalanceStatusBuilder()
+                                .withOptimizationResult(response.getJson().getJsonObject(CC_REST_API_SUMMARY).getMap())
+                                .addNewCondition().withNewType(ready.toString()).endCondition().build();
+                    } else {
+                        throw new RuntimeException("Rebalance returned unknown response: " + response.toString());
+                    }
+                }).otherwise(t -> {
+                    throw new RuntimeException(t.getMessage());
+                });
+    }
+
+    private RebalanceAnnotation getRebalanceAnnotation(KafkaClusterRebalance clusterRebalance) {
```

Comment:
Some Javadocs would not hurt ... you have `getRebalanceAnnotation`, `rebalanceAnnotation`, hasRebalanceAnnotation` ... not exactly clear what the difference is between those.